REF: add basic implementation of change signature refactoring

Kobzol opened the pull request:
This PR adds the `Change signature` refactoring.

![change-signature](https://user-images.githubusercontent.com/4539057/101150376-74b7fb80-3620-11eb-9c69-b0bf671a815c.gif)

The current implementation has basic functionality.
- ~How should the user be able to enter the name and type of parameters? I suggest that we use `RsPat` for name and `Ty` or `RsTypeReference` for types instead of plain strings, so that after the user enters something, we try to parse it and if it fails, we will not allow the dialog to proceed.~
- ~Currently, if a new parameter is added, it will not be filled with a default value at call sites, so that the user will not forget to update the call sites. This mirrors the behaviour of Kotlin.~
- ~Should the dialog offer things like `async`, `unsafe`, `extern` or variadic args? If yes, we need to let the user enter any possible value for these attributes. I think that it's worth to do it only if we can actually refactor something after changing them. Maybe removing/adding await/unsafe from call sites?~
- ~Originally I wanted to unify the Change signature and Extract function UI (and some data structures), but I realized that they actually differ quite a lot and it would complicate the code too much I think. I checked the Kotlin plugin and it has two different dialogs for these refactorings, so I think that it's ok.~
- ~I wanted to move all surrounding whitespace and comments when a parameter is moved from one place to another, but I'm having a difficulty making it work. The commented code in `insertItemWithComma` didn't work, basically if I try to do something like this:~
```kotlin
val inserted = parent.addBefore(<some whitespace>, anchor)
```
~the `inserted` will contain `anchor` instead of the inserted whitespace. This breaks copying of successive whitespace and comment elements, which I tried to do in a loop. Any hints of what might be wrong here?~

I'm not a huge fan of the signature config being mutable, but it makes UI and test code simpler and it also helps with detection of modified parameters (so I don't have to generate a GUID or something like that for them).

~The logic about detecting moved/removed/added parameters is slightly convoluted, it could be probably simplified, but I want to leave that for a final refactoring. It's not as simple as it sounds, as we must detect what parameters were removed, what were added and what were just moved, these operations have different semantics, even if they end up with the same result in the signature (see `test move and add parameter` and `test swap parameters`).~

Fixes: https://github.com/intellij-rust/intellij-rust/issues/4927

changelog: Add the Change Signature refactoring for functions and methods.

Kobzol commented:
@ortem @vlad20012 any feedback about my questions? :) Should I first create some GUI before we resolve them?

ortem commented:
Hi @Kobzol, sorry for the long delay!

> How should the user be able to enter the name and type of parameters? <...>

That's a good idea, probably `RsNamesValidator.kt` would help


> Currently, if a new parameter is added, it will not be filled with a default value at call sites, so that the user will not forget to update the call sites <...>

I don't have a strong opinion on this question. Do not fill arguments at the call site is generally OK for me, but if we decide to fill them, `org.rust.lang.core.psi.RsDefaultValueBuilder` may be helpful.

> Should the dialog offer things like async, unsafe, extern or variadic args? <...>

Basically, it would be great to support such things, including removing/adding something at the call site. But if will complicate this PR a lot, I would prefer merging a basic implementation without async/unsafe/extern support at first and try to implement it later. This PR is already quite big :)


> Originally I wanted to unify the Change signature and Extract function UI (and some data structures), but I realized that they actually differ quite a lot and it would complicate the code too much I think <...>

I think that's fine.


>I wanted to move all surrounding whitespace and comments when a parameter is moved from one place to another, but I'm having a difficulty making it work <...>

This happens because `com.intellij.psi.impl.source.tree.CompositeElement#addInternal` returns the first element that was appended skipping whitespaces. I'm not sure how we should deal with it. Maybe concatenate all whitespaces into one and add it just once. @vlad20012 do you have any ideas?

Kobzol commented:
I added basic support for `async` and `unsafe`. I also started working on the UI, however, I'm not sure what's the best approach to build the UI for the parameter table.

`RsExtractFunctionHandler` uses `AbstractParameterTablePanel`, which works fine, but I didn't find how to add the option for the user to remove and add parameters to it.

I checked the Kotlin plugin and it is using `ParameterTableModelBase`. I tried to use it too, but then I realized that it doesn't come with GUI (?). Do I have to build a custom UI from scratch for this?

ortem commented:
@Kobzol Have you checked the [Java implementation](https://github.com/JetBrains/intellij-community/blob/master/java/java-impl/src/com/intellij/refactoring/changeSignature/JavaChangeSignatureDialog.java)? It uses `JavaParameterTableModel` which extends `ParameterTableModelBase`

I'm currently not sure how to properly implement this dialog. But I'll try to investigate if the Java implementation doesn't help.

Kobzol commented:
I tried to use `ParameterTableModelBase`, it works, but it seems that it "only" provides the table model, but not any built-in GUI. I will build the GUI manually if necessary, but I wanted to check first if there is not a better way :)

Kobzol commented:
@ortem Thank you for the hint, I will probably use the Java built-in `Change Signature` dialog. It will need some hacks, but it should be hopefully easier than patching the parameter table into a custom dialog.

Kobzol commented:
@ortem I rewrote the UI using the Java dialog base class. It's far from perfect, both function-wise and UI-wise, but I think that it's now ready for a first round of review.

ortem commented on src/main/kotlin/org/rust/lang/core/psi/ext/RsFunction.kt:
```diff
@@ -172,6 +175,19 @@ val QueryAttributes.isProcMacroDef
+        ReferencesSearch.search(this, scope)
+    }
+    return search
+        .mapping { it.element.parentOfTypes(RsCallExpr::class, RsDotExpr::class) }
```
BTW Kotlin collections stuff can be used here instead of `Query` methods, e.g. `mapNotNull` and `filter`

ortem replied:
Just an idea, not sure how's better: introduce two separate methods to find all the `RsCallExpr` and all the  `RsMethodCall` (`RsDotExpr` is not needed here IIUC). Consider something like:
```Kotlin
fun RsFunction.findFunctionCalls(scope: SearchScope? = null): Sequence<RsCallExpr> {}
fun RsFunction.findMethodCalls(scope: SearchScope? = null): Sequence<RsMethodCall> {}
```

Kobzol replied:
Thank you, I'm sure that the call API will be useful to other parts of the code (for example the change signature refactoring :) ), so I introduced it in #6515.

ortem commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt:
```diff
@@ -0,0 +1,143 @@
+import org.rust.lang.core.types.type
+
+/**
+ * This class just holds `config`, otherwise it is unimplemented.
+ * It is required by ChangeSignatureProcessorBase.
```
```suggestion
 * This class just holds [config], otherwise it is unimplemented.
 * It is required by [com.intellij.refactoring.changeSignature.ChangeSignatureProcessorBase].
```

ortem commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt:
```diff
@@ -0,0 +1,143 @@
+ */
+class RsFunctionSignatureConfig private constructor(
+    val function: RsFunction,
+    var name: String = "",
```
```suggestion
    var name: String,
```

ortem commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt:
```diff
@@ -0,0 +1,143 @@
+    companion object {
+        fun create(function: RsFunction): RsFunctionSignatureConfig {
+            return RsFunctionSignatureConfig(function,
+                function.name ?: "",
```
```suggestion
                function.name.orEmpty(),
```

ortem commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,342 @@
+
+    private fun createNewParameter(descriptor: SignatureDescriptor): Parameter {
+        val pat = factory.createPat("p${descriptor.parametersCount}")
+        return Parameter(pat, factory.createType("()").type, NEW_PARAMETER)
```
```suggestion
        return Parameter(pat, factory.createType("()").type)
```

ortem commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt:
```diff
@@ -0,0 +1,143 @@
+            return RsFunctionSignatureConfig(function,
+                function.name ?: "",
+                function.valueParameters.mapIndexed { index, it ->
+                    Parameter(it.pat ?: RsPsiFactory(function.project).createPat("_"), it.typeReference?.type
```
Currently, the whole function body is hard to read. Consider some reformatting, probably like that
```Kotlin
val psiFactory = RsPsiFactory(function.project)
val parameters = function.valueParameters.mapIndexed { index, parameter ->
    Parameter(parameter.pat ?: psiFactory.createPat("_"), parameter.typeReference?.type ?: TyUnknown, index)
}
return RsFunctionSignatureConfig(
    function,
    function.name.orEmpty(),
    parameters,
    function.returnType,
    function.vis,
    function.isAsync,
    function.isUnsafe
)
```

ortem commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,342 @@
+        val mock = MOCK ?: error("You should set mock UI via `withMockChangeFunctionSignature`")
+        mock(config)
+        RsChangeSignatureProcessor(project, config.createChangeInfo()).run()
+
```
```suggestion
```

dima74 reviewed (commented):
> Currently, if a new parameter is added, it will not be filled with a default value at call sites, so that the user will not forget to update the call sites. This mirrors the behaviour of Kotlin.

Looks like Kotlin, C++, TypeScript and Python have ability to specify value for new parameter which should be used at call sites:

<img src="https://user-images.githubusercontent.com/6505554/103267489-aecbb180-49c2-11eb-9ccf-178555c4fe6e.png" width="50%">

> Should the dialog offer things like async, unsafe, extern or variadic args? <...>

I propose to add them in separate PR. Maybe even first add intention to add/remove `async`. Btw we already have intention to add `unsafe` if function has `unsafe` block

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,342 @@
+
+    override fun getVisibility(): String = ""
+
+    /**
+     * This needs to be false, because the default dialog only offers combo boxes for visibility, but we need
+     * arbitrary strings.
+     */
+    override fun canChangeVisibility(): Boolean = false
```
The problem is that the visibility of the function that is being refactored might not be one of these four values. Therefore, if I started to change the signature of e.g. `pub(in super::foo) fn foo()`, I could not build the dialog in such a way that it would represent the current state of the function.

In other words, if I used `Change signature`, then e.g. changed the name and executed the refactoring, it would change the visibility. Which is what I wanted to avoid (even though I agree that a combobox would be much nicer.

dima74 replied:
Maybe it is possible to create combobox with those variants plus current visibility of function (if it not covered by variants)?

Kobzol replied:
That would be a possible solution, yes. Is there something like a combobox combined with a text field available in the platform/plugin code already? That would be even better.

dima74 replied:
Maybe [`com.intellij.openapi.ui.ComboBox`](https://jetbrains.github.io/ui/controls/combo_box/) with `comboBox.setEditable(true)`

Kobzol replied:
Sigh, I swear that creating this small custom UI takes more time than half of the whole refactoring :D

Kobzol replied:
I used the editable combobox. Do you know if there is a way to embed a Rust code fragment into it, to get visibility auto-completion?

I can embed it into `EditorTextField`, because it has a notion of the IntelliJ `Document`, but the editor of the combobox is a Swing component that of course doesn't know anything about IntelliJ Documents.

dima74 replied:
Oh, I don't know how to do this. I think it is ok to not have autocompletion there, because complex restricted visibility (not `pub(super)` or `pub(crate)`) are extremely rare

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt:
```diff
@@ -0,0 +1,143 @@
+        fun create(function: RsFunction): RsFunctionSignatureConfig {
+            return RsFunctionSignatureConfig(function,
+                function.name ?: "",
+                function.valueParameters.mapIndexed { index, it ->
```
I wanted to do that, but I wonder how to provide a better error message? When I use `ChangeSignatureHandler`, it only has `findTargetMember`, which can return null, and in that case  `getTargetNotFoundMessage` is displayed.

So it would need to have a generic error message, something like `Must target a function with non-cfg disabled parameters`. Is it ok or is there a better way?

dima74 replied:
You can check it e.g. in `showRefactoringDialog` and if necessary show error message using `CommonRefactoringUtil.showErrorHint` (and not show dialog). Python [does this](https://github.com/JetBrains/intellij-community/blob/7d654ffc0eaee2129850b583112f235792671e75/python/src/com/jetbrains/python/refactoring/changeSignature/PyChangeSignatureHandler.java#L91-L103)

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureHandler.kt:
```diff
@@ -0,0 +1,342 @@
+        val function = config.function
+        val project = function.project
+
+        if (function.owner is RsAbstractableOwner.Trait) {
+            function.searchForImplementations().forEach {
+                val overridden = (it as? RsFunction) ?: return@forEach
+                changeSignature(project, config, overridden)
+            }
+        }
```
You're right, it's in my TODO list. Maybe for a future PR? To keep this one manageable.

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,342 @@
+        return fragment
+    }
+
+    override fun createCallerChooser(title: String?,
+                                     treeToReuse: Tree?,
+                                     callback: Consumer<MutableSet<RsFunction>>?): CallerChooserBase<RsFunction>? = null
```
You're right that this is the correct format. I don't like the format that I used here, but IntelliJ has started (maybe from 2020.3?) to format the parameters in such a way (when you use `Put parameters on separate lines`), so I have used it.

dima74 replied:
#6575

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,342 @@
+    }
+}
+
+class SignatureParameter(val factory: RsPsiFactory, val parameter: Parameter) : ParameterInfo {
```
Sadly they can't, since `ChangeSignatureDialog` is public and it uses these types as generic parameters.

dima74 replied:
Maybe we can also make `ChangeSignatureDialog` private

Kobzol commented:
> Looks like Kotlin, C++, TypeScript and Python have ability to specify value for new parameter which should be used at call sites:

Oh, I thought that the default value in Kotlin will do something like `fun test(a: Int = 5)`, but it indeed sets the default value at call sites. I can't say that I like it very much (the default value might differ amongst different call sites), but I suppose that I can add it. Probably should be left for another PR? This one is already quite big.

Kobzol commented:
I changed some small things and rebased onto master with the changes from https://github.com/intellij-rust/intellij-rust/pull/6515. After that I introduced several small commits for (hopefully) easier review. I'll try to send the rest tomorrow.

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureHandler.kt:
```diff
@@ -0,0 +1,342 @@
+                    if (parameter.pat is RsPatIdent && currentParameter.pat is RsPatIdent) {
+                        renameParameter(function.project, currentParameter, parameter.patText)
+                    } else {
+                        currentParameter.pat?.replace(parameter.pat)
```
Yes, since we allow arbitrary patterns, the refactoring can break code, but I don't think that we can do anything about that in general.

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureHandler.kt:
```diff
@@ -0,0 +1,342 @@
+    }
+}
+
+private fun changeArguments(factory: RsPsiFactory, usage: RsElement, parameterOps: List<ParameterOperation>) {
```
I wanted to avoid that because I thought that it would be difficult to reconstruct the original list (its whitespace, comments etc.). In other intentions/quick fixes/refactoring I was usually creating things from scratch and @Undin hinted to me that changing existing things should be generally a bit more stable.

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureHandler.kt:
```diff
@@ -0,0 +1,342 @@
+
+class RsChangeSignatureProcessor(project: Project, changeInfo: ChangeInfo)
+    : ChangeSignatureProcessorBase(project, changeInfo) {
+    override fun performRefactoring(usages: Array<out UsageInfo>?) {
```
I implemented the `findUsages` method and it seems that it is enough to display preview.

dima74 replied:
Looks like [all other languages implement `ChangeSignatureUsageProcessor`](https://user-images.githubusercontent.com/6505554/104351149-15e67b80-5516-11eb-9d42-e1454599d871.png), but I don't know whether or why it is needed.

Btw if I change function name (using preview to show usages) and then undo change signature refactoring, then dialog without function name can appear:
![image](https://user-images.githubusercontent.com/6505554/104351505-88575b80-5516-11eb-90ce-f67a31ae2f50.png)

Kobzol replied:
Ok, I'll implement it.

Kobzol replied:
I implemented it, but the undo button still says `Changing signature of ?`. I couldn't find from where the error message comes, any ideas?

dima74 replied:
Oh, actually it is because [`RsFindUsagesProvider#getDescriptiveName`](https://github.com/intellij-rust/intellij-rust/blob/b0be6aaeff0295cda97016507a0bb642d76b11a5/src/main/kotlin/org/rust/ide/search/RsFindUsagesProvider.kt#L23) always returns empty string

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureHandler.kt:
```diff
@@ -0,0 +1,342 @@
+                val currentParameter = parameters.valueParameterList[index]
+                if (parameter.pat.text != currentParameter.pat?.text) {
+                    if (parameter.pat is RsPatIdent && currentParameter.pat is RsPatIdent) {
+                        renameParameter(function.project, currentParameter, parameter.patText)
```
Hmm, I'm not sure how to unify these two. Arguments are changed differently than parameters. For arguments basically I just remove existing arguments or add empty commas. For parameters I need to shuffle them around, rename them/change their type or add new ones at the end.

dima74 replied:
I suggest to refactor `changeParameters` and `changeArguments` methods [like this](https://github.com/intellij-rust/intellij-rust/commit/7da38ed098062fd24cbd8ba6529bca4f4081bacb). That way:
* There will be common method `fixParametersOrder` (so no logic duplication for reordering parameters)
* We will use a bit more functional approach
* Code is shorter (about 100 lines)

Kobzol replied:
Thank you. I thought that it would be too complex to unify the logic, but 7 parameters (of `fixParametersOrder`) isn't that bad :D It looks much cleaner now, I just fixed removing all parameters of methods and modified the behaviour of trailing commas - I suppose that if we keep the trailing comma for multiple parameters, we should also keep them if there is only a single parameter?

Also, there is one test that I'm not sure how to make work.
```kotlin
fun `test add multiple parameters`() = doTest("""
    fn foo/*caret*/(a: u32) {}
    fn bar() {
        foo(0);
    }
""", """
    fn foo(a: u32, b: u32, c: u32) {}
    fn bar() {
        foo(0,,);
    }
""") {
    parameters.add(parameter("b", "u32"))
    parameters.add(parameter("c", "u32"))
}
```
The new code creates a different PSI than the PSI parsed from the test. So I'm not sure how to test this (except for changing the behaviour of the refactoring).

dima74 replied:
> I just fixed removing all parameters of methods

👍

> I suppose that if we keep the trailing comma for multiple parameters, we should also keep them if there is only a single parameter?

Looks like both intellij builtin formatter and rustfmt will anyway remove trailing comma if method is single line (and we convert method with one parameter to single line), so I think we should also remove it

> The new code creates a different PSI than the PSI parsed from the test

Yeah, the root cause is that `foo(1, , )` is parsed not as we expected (not as single `RsCallExpr`). We can workaround this by not adding whitespace between commas, so that auto reformatter will add whitespace and fix PSI on the way. That is change `{ factory.createWhitespace(" ") }` to `{ null }` in `changeArguments`

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureHandler.kt:
```diff
@@ -0,0 +1,342 @@
+
+private fun renameParameter(project: Project, parameter: RsValueParameter, newName: String) {
+    val binding = parameter.pat?.findBinding() ?: return
+    // TODO: how to run a refactoring inside another refactoring?
+    RefactoringFactory.getInstance(project).createRename(binding, newName).run()
```
`RenameUtil.rename` didn't work for me (it didn't rename the parameter usage), but I suppose that changing the RsPath manually also works.

Kobzol commented:
Usages are now found asynchronously using `findUsages`, which also makes preview work. Added better support for methods and improved visibility handling (it's not possible anymore to change the visibility of trait/impl trait methods and the validity of function visibility is now checked).

dima74 commented:
Some suggestions regarding dialog UI:
* Probably we don't need labels for async and unsafe checkboxes (checkboxes already have same text)
* Probably async and unsafe checkboxes can be on same row
* What width should have `return type`, `visibility` and `name` text fields? Maybe one third each
* Ideally space between `return type` and `visibility` fields should equal to space between `visibility` and `name` fields
* Which field should have focus initially? Currently it is `return type`, I would vote for parameters table
* When adding new parameter, user first have to delete default type `()`, and then enter desired type. Maybe we should use empty string as type for new parameter (if user doesn't change it, we will use `()`)

![image](https://user-images.githubusercontent.com/6505554/104352095-4a0e6c00-5517-11eb-85e0-1e435aae8a6e.png)

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,370 @@
+    }
+}
+
+private fun createVisibilityHints(initialVis: RsVis?): Array<String> {
+    val options = mutableListOf(
+        "",
+        "pub",
+        "pub(crate)"
+    )
+    val initialText = initialVis?.text.orEmpty()
+    if (initialText !in options) {
+        options.add(0, initialText)
+    }
+
+    return options.toTypedArray()
+}
```
We can't just use a set, because it wouldn't have a defined order. It's probably not ideal if the order of the options was different everytime the dialog was opened.

ortem replied:
@Kobzol This should be OK: [`setOf`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/set-of.html) uses `java.util.LinkedHashSet` so it preserves its arguments iteration order

Kobzol replied:
Oh, I didn't know that `setOf` guarantees it. Thank you :)

Kobzol commented:
I changed the order to `<visibility> <name> <return type>` to match the order in Rust code. This also solved the horizontal gap mismatch between the elements. The initial focus actually already is on the parameter table, but there there has to be at least a single parameter. Otherwise it will be set to the function name, which seems reasonable to me.

> When adding new parameter, user first have to delete default type (), and then enter desired type. Maybe we should use empty string as type for new parameter (if user doesn't change it, we will use ())

I'm not sure about this one. The type is represented in the parameter with `Ty`, so we can't easily represent an empty type (without resorting to hacks). We could either:
1) Display the `()` type as an empty string (but that could be confusing).
2) Store something like `Ty?` in the parameter and format `null` as an empty string.
3) Clear the existing type when the user clicks on it (or maybe do it only for `()`? or only the first time the user clicks on the parameter type?). Would probably be confusing.

dima74 commented:
> Store something like `Ty?` in the parameter and format `null` as an empty string.

I like this option

dima74 commented:
Some bugs related to lifetimes:

* Changing lifetime has no effect:

https://user-images.githubusercontent.com/6505554/104440914-031a8800-55a4-11eb-82f1-92e2438be2b8.mp4

* Changing type with lifetime removes lifetime:

https://user-images.githubusercontent.com/6505554/104441129-57be0300-55a4-11eb-80da-f72df7c01f96.mp4

dima74 commented:
* If user enters incorrect return type, it gets silently replaced to `_`. Probably we should show some error in this case
* If user enters parameter type with error, then it (or its part) gets replaced with `<unknown>`. Probably it is not good if e.g. there is big complex type, and user makes a typo, then that complex type will be lost. Maybe we shouldn't replace incorrect type with `<unknown>`, and show error when applying refactoring if parameter type is invalid.

https://user-images.githubusercontent.com/6505554/104442596-57bf0280-55a6-11eb-8248-5e1ebfdd4b88.mp4

* Also I think it is a bit unintuitive that some fields (parameter name and type) might change after they lose focus. Consider user enters incorrect type, and immediately apply refactoring ⇒ parameter type gets replaced to `_`:

https://user-images.githubusercontent.com/6505554/104443722-d8323300-55a7-11eb-9e20-9606a4ad1b3e.mp4

Kobzol commented:
Hmm, in light of this, it seems that `Ty` is not ideal, as it doesn't preserve e.g. lifetimes exactly. I'll try to switch it to `RsTypeReference`.

Kobzol commented:
I modified parameters so that invalid names and types are not auto-fixed on focus loss. The dialog now also displays errors properly and disables the refactoring buttons if any errors are found. I also added support for changing base trait methods, basic name and visibility conflict detection, implemented proper usage processing and finally fixed handling of whitespace and comments when modifying parameters and arguments (albeit at the cost of a pretty complex implementation :sweat_smile:).

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/impl.kt:
```diff
@@ -0,0 +1,405 @@
+    class MethodImplementation(val overriddenMethod: RsFunction) : RsFunctionUsage(overriddenMethod)
+}
+
+fun findFunctionUsages(function: RsFunction): List<RsFunctionUsage> {
+    return function.findUsages().map {
+        when (it) {
+            is RsCallExpr -> RsFunctionUsage.FunctionCall(it)
+            is RsMethodCall -> RsFunctionUsage.MethodCall(it)
+            is RsPath -> RsFunctionUsage.Reference(it)
+            else -> error("unreachable")
+        }
+    }.toList()
+}
```
```suggestion
fun findFunctionUsages(function: RsFunction): Sequence<RsFunctionUsage> {
    return function.findUsages().map {
        when (it) {
            is RsCallExpr -> RsFunctionUsage.FunctionCall(it)
            is RsMethodCall -> RsFunctionUsage.MethodCall(it)
            is RsPath -> RsFunctionUsage.Reference(it)
            else -> error("unreachable")
        }
    }
}
```

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureUsageProcessor.kt:
```diff
@@ -0,0 +1,141 @@
+    }
+    for (item in items) {
+        if (item == function) continue
+        if (item.name == config.name) {
```
Btw there will be a false-positive in case of multiple functions with same name but different cfg-attributes:
```rust
#[cfg(unix)]
fn foo() {}
#[cfg(windows)]
fn foo() {}
```
But I think it can be ignored for now

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureUsageProcessor.kt:
```diff
@@ -0,0 +1,141 @@
+            processFunctionUsage(config, usage)
+        }
+
+        return false
```
```suggestion
        return true
```
`ChangeSignatureProcessorBase` expects that `true` is returned on success (though it will not change anything because there is just single `ChangeSignatureUsageProcessor`)

Kobzol replied:
Oh, ok. I thought that I should let it process other processors if there would be more in the future.

dima74 replied:
Any other languages have single processor, so looks like it is unlikely that we will have more in future. Also please return `true` from `processPrimaryMethod` for same reason

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,389 @@
+    }
+
+    private fun createNewParameter(descriptor: SignatureDescriptor): Parameter =
+        Parameter(RsPsiFactory(descriptor.function.project), "p${descriptor.parametersCount}")
```
```suggestion
        Parameter(factory, "p${descriptor.parametersCount}")
```

dima74 commented on src/test/kotlin/org/rust/ide/refactoring/RsChangeSignatureTest.kt:
```diff
@@ -0,0 +1,885 @@
+    private fun createType(text: String): RsTypeReference = RsPsiFactory(project).createType(text)
+    private fun parameter(patText: String, type: RsTypeReference): Parameter {
+        val factory = RsPsiFactory(project)
+        return Parameter(
+            factory,
+            patText,
+            type
+        )
```
```suggestion
        return Parameter(factory, patText, type)
```

dima74 commented on src/test/kotlin/org/rust/ide/refactoring/RsChangeSignatureTest.kt:
```diff
@@ -0,0 +1,885 @@
+    """, """
+        fn foo(a: u32, b: u32) {}
+        fn bar() {
+            foo(0, );
```
Maybe we should add `todo!()` in function calls for added parameters?

Kobzol replied:
In my opinion it is better to notify the user about the invalid call sites immediately (by making the code fail to compile), so that they can be fixed ASAP. I wouldn't want to only find about it later at runtime. But I suppose that this can be solved with the default parameter values, if we implement them later? We can default these to `todo!()` possibly.

dima74 replied:
> it is better to notify the user about the invalid call sites immediately (by making the code fail to compile), so that they can be fixed ASAP

Yes, it is reasonable, so let's keep current behaviour

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,389 @@
+        }
+
+        return panel {
+            row {
+                asyncBox()
+                unsafeBox()
+            }
```
We should find a way to center checkboxes vertically:

![image](https://user-images.githubusercontent.com/6505554/105079119-08923980-5aa0-11eb-81f1-46b2f6bdde50.png)

And also ideally add some padding between checkboxes and help icon

Kobzol replied:
I can create a gap with `.withLargeLeftGap()`, but I can't seem to figure out the vertical alignment. Any hints? There doesn't seem to be a lot of (documented) layout API available for the Kotlin UI.

We can also put them in a column:
![image](https://user-images.githubusercontent.com/4539057/105089735-cb2cac80-5a9d-11eb-8cfd-78b7539bd91f.png)

dima74 replied:
Please try something like:
```kotlin
        return JPanel().apply {
            layout = MigLayout("align center center, insets 0 ${JBUI.scale(20)} 0 0")
            add(asyncBox)
            add(unsafeBox)
        }
```

Kobzol replied:
Oh, right, I don't have to use the UI DSL. Thanks for `MigLayout`, I didn't know about that.

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,389 @@
+                return "Parameter $index has invalid pattern"
+            }
+            if (parameter.typeText == null) {
+                return "Parameter $index has invalid type"
```
Hmm, I'm not sure what to do about the second problem. The update function is only fired after you focus out of the type code fragment or after you write some character (then it's updated with the previous value).

dima74 replied:
> The update function is only fired after you focus out of the type code fragment or after you write some character (then it's updated with the previous value)

Maybe when validating we should directly use code fragment text (instead of our variable which gets updated in callback)

Kobzol replied:
I'm using `ParameterTableModelBase.TypeColumn`, which requires a code fragment as a value. I can also try creating a new column type from scratch.

Kobzol replied:
Oh, you're right. I also had to change it for the signature, but it works. Thanks.

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/impl.kt:
```diff
@@ -0,0 +1,405 @@
+                val newParameter = factory.createValueParameter(parameter.patText, typeReference, reference = false)
+                val anchor = findAnchorToInsertItem(parameters, function, index)
+                insertItemWithComma(factory, newParameter, parameters, anchor)
+                RsImportHelper.importTypeReferencesFromTy(function, parameter.type, useAliases = true)
```
I tried that originally, but the import didn't work with the type reference, only with the type. Now that I have switched the parameter type to a String, the import doesn't work at all. For example, in this test:
```rust
mod foo {
    pub struct S;
             //^
}
mod bar {
    fn baz/*caret*/() {}
}
```
Adding a parameter with type `S` creates a type reference which has `type == TyUnknown`, so no import happens even if `importTypeReferencesFromElements` is used.

dima74 replied:
> Now that I have switched the parameter type to a String, the import doesn't work at all.

Oh, so looks like we should switch back to `RsTypeReference` (but with rendering it as empty string if it is `null`)

> I tried that originally, but the import didn't work with the type reference, only with the type.

Actually I don't know how import can work, because if user accepts autocompletion, then import gets added even before refactoring was started, and otherwise type will indeed be unknown. Probably we should use some fresh file as context for type fields, and set original file as context for this fresh file (so any imports during dialog editing are added to fresh file and not original file)

https://user-images.githubusercontent.com/6505554/105155603-db836c80-5b1b-11eb-8ab9-56235c04ca5b.mp4

Kobzol replied:
> Oh, so looks like we should switch back to RsTypeReference (but with rendering it as empty string if it is null)

Well the type reference is only useful for tests, in GUI it will not refer to existing types anyway (unless they are auto-imported during editing). I'll try the separate file to see if it helps.

Kobzol replied:
The behaviour of importing something into a file from a code fragment is intended generally? Or should I change it for all code fragments?

dima74 replied:
> The behaviour of importing something into a file from a code fragment is intended generally?

* For code fragment used in debugger definitely not intended - #6699
* For other code fragments I don't know, but I suppose that e.g. for [code fragment used in macro completion](https://github.com/intellij-rust/intellij-rust/blob/0f5790ecbd2d152fe9265d4b3160b9749269a48b/src/main/kotlin/org/rust/lang/core/completion/RsPartialMacroArgumentCompletionProvider.kt#L61-L64) it is intended behaviour

> Or should I change it for all code fragments?

So if possible maybe add some flag to code fragment constructor (to control behaviour case-by-case)

Kobzol replied:
Right, that makes sense, thanks. So far I only managed to change the import location from the file to the fragment itself :D Which is of course not ideal. I suppose that to create a virtual file, I'd need to somehow connect it to the corresponding module structure, so that it sees the correct paths from the perspective of the function, and then direct the imports into it.

I guess that e.g. the new macro engine also needs to build something like that? @vlad20012 any hints as to what API to use?

dima74 replied:
We can try something similar to [used in repl console](https://github.com/intellij-rust/intellij-rust/blob/0f5790ecbd2d152fe9265d4b3160b9749269a48b/src/main/kotlin/org/rust/ide/console/RsConsoleCodeFragmentContext.kt#L103-L106) (I don't know whether it will work or not):
```kotlin
    val function: RsFunction = ...
    val freshFile = RsPsiFactory(project).createFile("fn main() {}")
    freshFile.originalFile = function.containingFile as RsFile
    return RsTypeReferenceCodeFragment(project, "", context = freshFile)
```

Kobzol replied:
Thanks for the hint, I included an `importTarget` in the code fragments and check it in `ImportCandidate.import`. It seems to work fine.

dima74 replied:
Looks like it works in tests, but didn't work in IDE...

https://user-images.githubusercontent.com/6505554/105376387-bc1f3900-5c1a-11eb-9a5d-48b2c4598b14.mp4

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt:
```diff
@@ -0,0 +1,134 @@
+) {
+    val type: Ty
+        get() = typeText?.type ?: TyUnit
+    val typeReference: RsTypeReference
+        get() = typeText ?: factory.createType("()")
```
I changed it now to an empty string. It's a bit annoying that immediately after creating a parameter, the dialog will display an error (`Please fill the parameter type`), but the Koltin plugin does the same, so probably it's fine.

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,389 @@
+    : ParameterTableModelBase<SignatureParameter, ModelItem>(
+    descriptor.function,
+    descriptor.function,
+    NameColumn<SignatureParameter, ModelItem>(descriptor.function.project, "Pattern:"),
```
I added it for consistency with `Type:`. But I can also change `Type:` to `Type` :)

dima74 replied:
I would say that in Kotlin these labels are usual labels for text fields (and not table header labels), but looks like it is not that important, so please do as you like :)

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureUsageProcessor.kt:
```diff
@@ -0,0 +1,141 @@
+        if (beforeMethodChange) return false
+
+        val rsChangeInfo = changeInfo as? RsSignatureChangeInfo ?: return false
+        val usage = usageInfo as? RsFunctionUsage ?: return false
```
Please also change three other patterns with `rsChangeInfo` (at least to have consistent style)

dima74 commented on src/test/kotlin/org/rust/ide/refactoring/RsChangeSignatureTest.kt:
```diff
@@ -0,0 +1,893 @@
+            foo(0);
+        }
+    """) {
+        parameters.removeAt(parameters.size - 1)
```
```suggestion
        parameters.removeLast()
```

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt:
```diff
@@ -90,14 +90,13 @@ class RsChangeFunctionSignatureConfig private constructor(
+    }
 
-    fun signature(): String = buildString {
+    fun signature(parameters: List<Parameter>): String = buildString {
```
This is.. not pretty, but the problem is that `calculateSignature()` seems to be called before `updateState` for some reason, so this seemed like the easiest solution.

Kobzol replied:
Nevermind, I solved it in a different way.

Kobzol commented:
> Looks like both intellij builtin formatter and rustfmt will anyway remove trailing comma if method is single line (and we convert method with one parameter to single line), so I think we should also remove it

That's true, but (at least `rustfmt`) also removes trailing comma if there are multiple parameters on a single line. So maybe the condition shouldn't care about the number of parameters, but rather about single/multi line?

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/impl.kt:
```diff
@@ -0,0 +1,293 @@
+
+    // add parameters one by one
+    val rspace = parametersCopy.lastChild.prevSibling as? PsiWhiteSpace
+    val isSingleParameter = descriptors.size == 1 && descriptors.size < config.originalParameters.size
```
```suggestion
    val isSingleParameter = groups.size == 1 && descriptors.size < config.originalParameters.size
```

Looks like I also forget about `self` parameter here. Also please add test for removing parameter in multiline method like:
```rust
impl Foo {
    fn foo(
        &self,
        a: i32,
        b: i32,
    ) {}
}
```

dima74 commented:
> That's true, but (at least `rustfmt`) also removes trailing comma if there are multiple parameters on a single line. So maybe the condition shouldn't care about the number of parameters, but rather about single/multi line?

Yes, that would be reasonable. So something like:

```rust
if (hasTrailingComma && isMultiline && !isSingleParameter) parameters.addBefore(factory.createComma(), rparen)
```

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/impl.kt:
```diff
@@ -0,0 +1,302 @@
+    // Keep trailing comma if the resulting parameter list will have at least two parameters on multiple lines
+    if (hasTrailingComma &&
+        !isSingleParameter &&
+        (descriptors.isNotEmpty() || parameterSelf != null) &&
```
```suggestion
```
this part is always true because of `if (descriptors.isEmpty() && parameterSelf == null) return` line above. Also maybe move this `if` inside else branch of `if (isSingleParameter) {`

Kobzol replied:
Sorry for this mess :D I originally used a different condition at the beginning of the function and forgot that it covers this scenario. So in the end the only `diff` is `if (hasTrailingComma)` -> `if (hasTrailingComma && isMultiline)` :man_facepalming:

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt:
```diff
@@ -0,0 +1,134 @@
+            val self = function.selfParameter?.text.orEmpty()
+            val prefix = if (self.isNotEmpty()) ", " else ""
+            val parameters = parameters.joinToString(", ", prefix=prefix) { "${it.pat.text}: ${it.typeReference.text}" }
+            return "${self}$parameters"
```
Looks like we need to fix it again :)

Kobzol replied:
Copy-paste strikes again.. Thank you.

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt:
```diff
@@ -0,0 +1,396 @@
+        }
+
+        for ((index, parameter) in config.parameters.withIndex()) {
+            if (parameter.parsePat() == null) {
+                return "Parameter $index has invalid pattern"
+            }
+            if (parameter.type is ParameterType.Empty) {
+                return "Please enter type for parameter $index"
+            }
+            if (parameter.type is ParameterType.Invalid) {
+                return "Type entered for parameter $index is invalid"
+            }
```
Wouldn't it be more convenient for user if we refer to parameter by name instead of by index?

Kobzol replied:
I wanted to do that, but the parameter might:
1) Not have a name (it might be empty)
2) It might be a complex pattern (e.g. `Struct { a: Struct2(x, y), b: Struct3(z), .. }`).

Therefore, I think that the index is more general. What do you think? Or maybe we can show the name if the pattern is non-empty and simple, and in other cases show the index.

dima74 replied:
Oh, indeed, there are so many possibilities. Then let's keep index

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt:
```diff
@@ -0,0 +1,164 @@
+    val typeReference: RsTypeReference
+        get() = parseTypeReference() ?: factory.createType("()")
+
+    fun parseTypeReference(): RsTypeReference? = when (val activeType = type) {
+        is ParameterType.Valid -> activeType.typeReference
+        else -> null
+    }
```
```suggestion
    fun parseTypeReference(): RsTypeReference? = (type as? ParameterType.Valid)?.typeReference
```

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt:
```diff
@@ -0,0 +1,164 @@
+    companion object {
+        fun create(function: RsFunction): RsChangeFunctionSignatureConfig {
+            val parameters = function.valueParameters.mapIndexed { index, parameter ->
+                Parameter(RsPsiFactory(function.project), parameter.pat?.text
+                    ?: "_", ParameterType.fromTypeReference(parameter.typeReference), index)
```
```suggestion
                val factory = RsPsiFactory(function.project)
                val patText = parameter.pat?.text ?: "_"
                val type = ParameterType.fromTypeReference(parameter.typeReference)
                Parameter(factory, patText, type, index)
```

Or at least, can we please have line break after comma, not in the middle of argument

dima74 commented on src/main/kotlin/org/rust/lang/core/psi/RsCodeFragment.kt:
```diff
@@ -123,8 +126,11 @@ class RsStatementCodeFragment(project: Project, text: CharSequence, context: RsE
 
-class RsTypeReferenceCodeFragment(project: Project, text: CharSequence, context: RsElement)
-    : RsCodeFragment(project, text, RsCodeFragmentElementType.TYPE_REF, context),
+class RsTypeReferenceCodeFragment(project: Project,
+                                  text: CharSequence,
+                                  context: RsElement,
+                                  importTarget: RsElement? = null)
+    : RsCodeFragment(project, text, RsCodeFragmentElementType.TYPE_REF, context, importTarget = importTarget),
```
```suggestion
class RsTypeReferenceCodeFragment(
    project: Project,
    text: CharSequence,
    context: RsElement,
    importTarget: RsElement? = null
) : RsCodeFragment(project, text, RsCodeFragmentElementType.TYPE_REF, context, importTarget = importTarget),
```

dima74 commented on src/main/kotlin/org/rust/ide/utils/import/importUtils.kt:
```diff
@@ -59,17 +59,20 @@ fun ImportCandidate.import(context: RsElement) {
+                ?: context.ancestors.findLast { it is RsFunction }
+            ((scope as? RsFunction)?.block ?: scope) as RsItemsOwner
+        }
+        containingFile is RsCodeFragment -> containingFile.importTarget as? RsItemsOwner
```
Maybe change `RsCodeFragment.importTarget` type to `RsItemsOwner` (since we can import only to it)

dima74 commented on src/main/kotlin/org/rust/lang/core/psi/RsCodeFragment.kt:
```diff
@@ -26,20 +26,23 @@ abstract class RsCodeFragment(
-    forceCachedPsi: Boolean = true
-) : RsFileBase(fileViewProvider), PsiCodeFragment {
+    forceCachedPsi: Boolean = true,
+    val importTarget: RsElement? = null
```
```suggestion
    val importTarget: RsElement? = null,
```

Kobzol replied:
I haven't seen this style applied elsewhere (e.g. `RsPathCodeFragment` in the same file), is trailing comma in multiline parameters the correct style?

dima74 replied:
Kotlin compiler started to support trailing comma only few months ago (somewhere in September), so it technically can't be used in `RsPathCodeFragment`. Also I am not aware of any consensus regarding the style, but it seems reasonable to use it in new code for better `git blame` (e.g. I [use](https://github.com/intellij-rust/intellij-rust/blob/44d8830e1c9f396137120a908d69bcb2387946b2/src/main/kotlin/org/rust/lang/core/resolve2/CrateDefMap.kt#L334-L337) it)

Kobzol replied:
Ah, now I remember, one of the most awaited features of Kotlin 1.4 :D Ok, I'll try to remember to use it. The default formatter doesn't apply it, which is a shame though, can we change it as you have recently changed the formatting of multiline parameters?.

dima74 replied:
We discussed and decided that it is not that important to think about it

Kobzol commented:
> Looks like it works in tests, but didn't work in IDE...

It will only work if you actually import the type (into the fake file), either by clicking on the type name from the completion list, or by using `Alt + Enter` -> Import quick fix. Otherwise the type will be unresolved. What we could do in theory is to import the type references into the fake file automatically just before running the refactoring, but if there were more import candidates, it would be problematic.

dima74 commented:
> It will only work if you actually import the type (into the fake file), either by clicking on the type name from the completion list, or by using `Alt + Enter` -> Import quick fix

Maybe I am doing something wrong, but for me it doesn't work in any of two cases

Kobzol commented:
Interesting. It seems that sometimes it doesn't work, I'll try to debug.
![import](https://user-images.githubusercontent.com/4539057/105389000-9cd2dc80-5c17-11eb-9c1d-efb5b0ecc65f.gif)

Kobzol commented:
Hmm, for some reason sometimes the code fragment can resolve the type, sometimes not. Even if the fake file has the import inserted, sometimes the fragment does not resolve the type. I tried to calling `clearCaches()` on the fragment, but I'm not sure if that makes sense. Is there some API to "clear caches" or force a new resolve cycle?

dima74 commented:
> Is there some API to "clear caches"

`project.rustPsiManager.incRustStructureModificationCount()`, but I am not sure that we should use it

Kobzol commented:
> project.rustPsiManager.incRustStructureModificationCount(), but I am not sure that we should use it

Anecdotally, with this call it seems to work.

dima74 commented:
> Anecdotally, with this call it seems to work.

So could you please push it

dima74 commented:
Also thanks a lot for independent commits, it was extremely helpful during review. Now when we are close to merging, could you please squash commits

Kobzol commented:
Sorry for the delay. Squashed commits and rebased over master to fix conflict.

dima74 commented on src/main/kotlin/org/rust/ide/refactoring/functionSignature.kt:
```diff
@@ -0,0 +1,35 @@
+        get() {
+            val typeParams = typeParameters()
+            if (typeParams.isEmpty()) return ""
+            return typeParams.joinToString(separator = ",", prefix = "<", postfix = ">") { it.text }
```
```suggestion
            return typeParams.joinToString(separator = ", ", prefix = "<", postfix = ">") { it.text }
```

Kobzol replied:
Should I also include the space after comma in `whereClausesText`?

dima74 replied:
I think yes, so it would be consistent with default style of IntelliJ formatter and rustfmt

dima74 reviewed (commented):
When adding parameter of type `HashSet`, import for `RandomState` is added:

https://user-images.githubusercontent.com/6505554/105693713-1b2dc800-5f11-11eb-8800-c8e16450a211.mp4

Kobzol commented:
> When adding parameter of type HashSet, import for RandomState is added:

This is a "problem" with `RsImportHelper`, which does not skip type arguments with default value. It's also present in other places (like change return type fix, substitute associated type intention, etc.). Maybe it should be fixed in a separate PR?

dima74 commented:
> Maybe it should be fixed in a separate PR?

Yes, it is good idea

dima74 commented on src/main/kotlin/org/rust/lang/core/psi/RsPsiFactory.kt:
```diff
@@ -389,8 +393,14 @@ class RsPsiFactory(
+        reference: Boolean = true,
+        lifetime: RsLifetime? = null
+    ): RsValueParameter {
+        return createFromText<RsFunction>("fn main($name: ${if (reference) "&" else ""}${if (lifetime != null) lifetime.text + " " else ""}${if (mutable) "mut " else ""}${type.text}){}")
```
```suggestion
        val referenceText = if (reference) "&" else ""
        val lifetimeText = if (lifetime != null) lifetime.text + " " else ""
        val mutText = if (mutable) "mut " else ""
        return createFromText<RsFunction>("fn main($name: $referenceText$lifetimeText$mutText${type.text}){}")
```

dima74 reviewed (approved):
Thanks! 🎉🎉🎉
bors r+

Kobzol commented:
bors retry

ortem commented:
@Kobzol @dima74 Impressive work!
